src/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs

```diff
@@ -37,6 +37,108 @@ public static void InitialStateMultiSegment()
+                Assert.Equal(JsonCommentHandling.Disallow, json.CurrentState.Options.CommentHandling);
+
+                json.Read(); // this should throw
+            });
+        }
+
+        [Theory]
+        [InlineData("2e2", 200)]
+        [InlineData("2e+2", 200)]
+        [InlineData("123e-01", 12.3)]
+        [InlineData("0", 0)]
+        [InlineData("0.1", 0.1)]
+        [InlineData("-0", 0)]
+        [InlineData("-0.1", -0.1)]
+        [InlineData("   2e2   ", 200)]
+        [InlineData("   2e+2   ", 200)]
+        [InlineData("   123e-01   ", 12.3)]
+        [InlineData("   0   ", 0)]
+        [InlineData("   0.1   ", 0.1)]
+        [InlineData("   -0   ", 0)]
+        [InlineData("   -0.1   ", -0.1)]
+        [InlineData("[2e2]", 200)]
+        [InlineData("[2e+2]", 200)]
+        [InlineData("[123e-01]", 12.3)]
+        [InlineData("[0]", 0)]
+        [InlineData("[0.1]", 0.1)]
+        [InlineData("[-0]", 0)]
+        [InlineData("[-0.1]", -0.1)]
+        [InlineData("{\"foo\": 2e2}", 200)]
+        [InlineData("{\"foo\": 2e+2}", 200)]
+        [InlineData("{\"foo\": 123e-01}", 12.3)]
+        [InlineData("{\"foo\": 0}", 0)]
+        [InlineData("{\"foo\": 0.1}", 0.1)]
+        [InlineData("{\"foo\": -0}", 0)]
+        [InlineData("{\"foo\": -0.1}", -0.1)]
+        public static void ReadJsonWithNumberVariousSegmentSizes(string input, double expectedValue)
+        {
+            byte[] utf8 = Encoding.UTF8.GetBytes(input);
+
+            var jsonReader = new Utf8JsonReader(utf8);
+            ReadDoubleHelper(ref jsonReader, utf8.Length, expectedValue);
+
+            ReadOnlySequence<byte> sequence = JsonTestHelper.GetSequence(utf8, 1);
+            jsonReader = new Utf8JsonReader(sequence);
+            ReadDoubleHelper(ref jsonReader, utf8.Length, expectedValue);
+
+            for (int splitLocation = 0; splitLocation < utf8.Length; splitLocation++)
+            {
+                sequence = JsonTestHelper.CreateSegments(utf8, splitLocation);
+                jsonReader = new Utf8JsonReader(sequence);
+                ReadDoubleHelper(ref jsonReader, utf8.Length, expectedValue);
+            }
+
+            for (int firstSplit = 0; firstSplit < utf8.Length; firstSplit++)
+            {
+                for (int secondSplit = firstSplit; secondSplit < utf8.Length; secondSplit++)
+                {
+                    sequence = JsonTestHelper.CreateSegments(utf8, firstSplit, secondSplit);
+                    jsonReader = new Utf8JsonReader(sequence);
+                    ReadDoubleHelper(ref jsonReader, utf8.Length, expectedValue);
+                }
+            }
+        }
+
+        private static void ReadDoubleHelper(ref Utf8JsonReader jsonReader, int expectedBytesConsumed, double expectedValue)
+        {
+            while (jsonReader.Read())
+            {
+                if (jsonReader.TokenType == JsonTokenType.Number)
+                {
+                    Assert.Equal(expectedValue, jsonReader.GetDouble());
```

Comment:
Is there anything tighter you can assert here?

```C#
Assert.Equals(jsonReader.GetValueSpanOrSegmentLength(), jsonReader.BytesConsumed - jsonReader.TokenStartIndex);
```

maybe?  (fake method as shorthand for the inevitable if)